src/ts_catalog/continuous_agg.c

```diff
@@ -1487,34 +1499,137 @@ ts_continuous_agg_bucket_width(const ContinuousAgg *agg)
 }
 
 /*
- * Determines bucket size in months for given bucketing function.
+ * Calls one of time_bukcket_ng() versions depending on the arguments. This is
+ * a common procedure used by ts_compute_* below.
+ */
+static Datum
+generic_time_bucket_ng(const ContinuousAggsBucketFunction *bf, Datum timestamp)
+{
+	if (strlen(bf->timezone) > 0)
+	{
+		return DirectFunctionCall3(ts_time_bucket_ng_timezone,
+								   IntervalPGetDatum(bf->bucket_width),
+								   timestamp,
+								   CStringGetTextDatum(bf->timezone));
+	}
+
+	return DirectFunctionCall2(ts_time_bucket_ng_timestamp,
+								   IntervalPGetDatum(bf->bucket_width),
+								   timestamp);
+}
+
+/*
+ * Adds one bf->bucket_size interval to the timestamp. This is a common
+ * procedure used by ts_compute_* below.
+ *
+ * If bf->timezone is specified, the math happens in this timezone.
+ * Otherwise, it happens in UTC.
+ */
+static Datum
+generic_add_interval(const ContinuousAggsBucketFunction *bf, Datum timestamp)
+{
+	Datum tzname = 0;
+	bool has_timezone = (strlen(bf->timezone) > 0);
+
+	if (has_timezone)
+	{
+		/*
+		 * Convert 'timestamp' to TIMESTAMP at given timezone.
+		 * The code is equal to 'timestamptz AT TIME ZONE tzname'.
+		 */
+		tzname = CStringGetTextDatum(bf->timezone);
+		timestamp = DirectFunctionCall2(timestamptz_zone, tzname, timestamp);
+	}
+
+	timestamp =
+		DirectFunctionCall2(timestamp_pl_interval, timestamp, IntervalPGetDatum(bf->bucket_width));
+
+	if (has_timezone)
+	{
+		Assert(tzname != 0);
+		timestamp = DirectFunctionCall2(timestamp_zone, tzname, TimestampGetDatum(timestamp));
+	}
+
+	return timestamp;
+}
+
+/*
+ * Computes inscribed refresh_window for variable-sized buckets.
+ *
+ * The algorithm is simple:
+ *
+ * end = time_bucket(bucket_size, end)
+ *
+ * if(start != time_bucket(bucket_size, start))
+ *     start = time_bucket(bucket_size, start) + interval bucket_size
+ *
+ */
+void
+ts_compute_inscribed_bucketed_refresh_window_variable(int64 *start, int64 *end,
+													  const ContinuousAggsBucketFunction *bf)
+{
+	Datum start_old, end_old, start_new, end_new;
+
+	start_old = ts_internal_to_time_value(*start, TIMESTAMPOID);
```

Comment:
It is non-obvious why hard-coding `TIMESTAMPOID` is OK here. There's nothing in the function interface that lets you know that this is an operation on start and end as a timestamp.


So, my question is if it is OK to implicitly assume `TIMESTAMP` type here? Should we pass the type as an argument to the function instead? At the very least, I think the function name should convey that we are computing this under the assumption that start and end is in a timestamp range.

Reply:
Yes, it's OK to use TIMESTAMPOID. I will add a comment.

Reply:
Here is the updated patch with added comments.